pxr/base/tf/fileUtils.cpp

```diff
@@ -94,6 +93,21 @@ Tf_HasAttribute(
         }
         return false;
     }
+
+    // Ignore reparse points on network volumes. They can't be resolved
+    // properly, so simply remove the reparse point attribute and treat
+    // if like a regular file/directory.
+    if ((attribs & FILE_ATTRIBUTE_REPARSE_POINT) != 0) {
+        // Calling PathIsNetworkPath sometimes sets the "last error" to an
+        // error code indicating an incomplete overlapped I/O function. We
+        // want to ignore this error.
+        DWORD olderr = GetLastError();
+        if (PathIsNetworkPathW(ArchWindowsUtf8ToUtf16(path).c_str())) {
+            attribs &= ~FILE_ATTRIBUTE_REPARSE_POINT;
```

Comment:
We're unconditionally altering attribs, regardless of the value of `resolveSymlinks`.

This means that if we were asking  "Hey, DON'T try to resolve symlinks, and just tell me if this path IS a symlink"... we'd now we'd be answering, "No, it isn't a symlink"... which feels wrong.

Also - in the case where we're just checking for file existence, the old code resolves symlinks recursively before providing an answer, which implies that is necessary, at least for non-network paths. For network paths, does it work differently? If not, we'd be answering that the file exists, when it may not (ie, there's a broken link, even on the local computer).  (In that case, we might need a different method to check file existence - some windows specific method for just checking existence, that handles reparse points automatically?)

Reply:
I think this is the correct thing to do here because other code is going to make decisions (such as altering the path string to point to the symlink location) based on the answer to the question "is this directory a symlink". The whole point of this change is that no code anywhere should treat reparse points on network drives as reparse points. So I intentionally completely obscure the fact that the reparse points on network drives even reparse points.

That's an interesting point about broken links and file existence. But here's the thing - prior to this change _any_ attempt to check the existence of a directory that was a reparse point on a network drive  would try to follow the symlink, not find it, and _always_ return false for the file existence of `S:\path\to\some`. Now, if we find a directory with a reparse point we always return true, which I think is totally valid because the requested directory does in fact exist. If the destination of the symlink doesn't exist does it have any children? No. Can you make a file inside it? No. But at least now you can try. And the whole basis for needing this PR is that just by looking at `S:\path\to\some` there is literally no way from the remote machine to determine if the destination of the reparse point exists. So even if we wanted to do what you suggest I don't think it's possible.

Reply:
tl;dr
-----
Overall, I think this PR is a good idea, and a move in the right direction, so I think it should be merged.

However, I do think we should at least add documentation to `Tf_HasAttribute`, `TfPathExists`, `TfIsDir`, `TfIsFile`, and `TfIsLink` explaining this otherwise hidden behavior - and possibly consider adding a `bool ignoreNetworkReparsePoints=true` parameter to them as well.

Also, in the case of existence checks with `resolveSymlinks=true`, I think we should use `FindFirstFile` as a way to validate if a network junction is "working".

But hopefully neither of those are particularly hard changes...?

Full Details
------------

> I think this is the correct thing to do here because other code is going to make decisions (such as altering the path string to point to the symlink location) based on the answer to the question "is this directory a symlink". The whole point of this change is that no code anywhere should treat reparse points on network drives as reparse points. So I intentionally completely obscure the fact that the reparse points on network drives even reparse points.

I see your point.  I think ignoring the "link nature" will give the behavior people probably want more often than not.

However - that assessment relies on some assumptions:
- People are mostly calling `TfIsLink` to know whether or not to call `TfReadLink`
- A result from `TfReadLink` that results in an inaccessible path will generally not be useful
- That people won't be using "true" windows symlinks (IO_REPARSE_TAG_SYMLINK) over a network share

I think all those assumptions will generally be true - but it's worth noting that there are some scenarios where each of them may not be:

- There are cases where you want to know if something is a link, and don't care what it's target is:
  - Some recursive crawls or copies may always avoid symlinks to prevent potential recursion or double-copying
  - You might record names and attributes (including whether it's a symlink) to see if they've changed - say, as part of a directory-diff tool.  You might expect such a diff generated locally and through a network share to be the same.
- There are cases where even an "inaccessible" path is useful:
  - You're working in a context with "known paths", and can do your own substitutions - ie, it's not hard to imagine that a VFX facility somewhere has written code that "knows" that `Z:\my_share` and `\\MYSERVER\zshare` are interchangeable
  - You are only interested in a portion of the resolved path - for instance, you might have a directory structure like this:
    ```
    Z:\shared\apps\FuzzyPuppyFactory
        v1.0.1\
        v1.3.2\
        v2.1.0\
        latest [junction: Z:\shared\apps\FuzzyPuppyFactory\v2.1.0]
    ```
     ...and you have code which calls `TfReadLink("\\AppServer\apps\FuzzyPuppyFactory\latest")` and grabs the last token to figure out the version
  - You want a "canonical version" of a path for comparison - ie, you don't care if you can't access `Z:\some\path\myfile.usd`, you just want to look to look it up in a database using that path as a key
- It is possible to use "true" windows symlinks over a network share.. it's just not possible by default.
  - To even create the symlinks, you need to either be an Administrator, or have the "Create symbolic links" privilege granted
  - To allow symbolic links to work over shares, the local computer must have "symlinkevaluation" property set to enable remote links, ie:
    - `fsutil behavior set SymlinkEvaluation R2R:1 R2L:1`
  - While that's a pain, it's certainly possible that in a managed environment, like a VFX studio, companies might have granted these permissions, and be relying on them

Again - I think those assumptions are generally true, so I think this change does more overall help than harm.  But the chance of breaking workflows is something we should be aware of.

Also, it does makes me a bit uncomfortable to do "hidden magic" like this.  In the situations where someone IS in one of the corner cases like above, it could result in a hard-to-track-down bug.  Can we at least add a clear comment to the docstrings (of both `Tf_HasAttribute`, and the public functions that use it - `TfPathExists`, `TfIsDir`, `TfIsFile`, and `TfIsLink`) explaining this behavior?  Or, we could go one step farther, and expose a parameter to control this behavior - say, `bool ignoreNetworkReparsePoints=true` - on the above functions.  That would both this behavior more discoverable, and provide a workaround if someone relies on old behavior.

> prior to this change any attempt to check the existence of a directory that was a reparse point on a network drive would try to follow the symlink, not find it, and always return false for the file existence of S:\path\to\some. Now, if we find a directory with a reparse point we always return true, which I think is totally valid because the requested directory does in fact exist.

I definitely agree that always returning false for existence of remote junctions is something that needs to be fixed - and is part of the reason why I think this change is overall a good thing.  However, always returning true for a network reparse point, even if `resolveSymlinks=true` and the link is actually broken, seems a step in the wrong direction.  The purpose of checking existence with `resolveSymlinks=true` is to get a better idea of "is this going to work". Again, I agree that universally returning true is an improvement over universally returning false - as it's more likely that a junction is going to work than not - or at least, I would hope so!  But it would be best if we could actually get a meaningful answer to the "will it work" question.

> And the whole basis for needing this PR is that just by looking at S:\path\to\some there is literally no way from the remote machine to determine if the destination of the reparse point exists. So even if we wanted to do what you suggest I don't think it's possible.

It's true that, if you were starting just from `S:\path\to\some`, it would be impossible to know if that works. But in our case, we can just do a heuristic test - ie, see if we can list the contents of the network path, `//B/share/path/to/some`.  It's true that there are situations where the directory could exist, but you can't list it's contents - permissions issues, for one - but following the "do more good than harm" guidelines, this feels like a move in the right direction.  I did a test with `FindFirstFile` with a network share, and it seemed to work as a reasonable heuristic for whether the link is broken or not.

Reply:
Could you clarify how you are using FindFirstFile? What if the directory exists but doesn't contain anything? Are you relying on the error code from FindFirstFile to distinguish between these two cases?

I'll preface all this by saying I think you and I may have different objectives here :) It matters enormously to me that the USD library works the way a user expects it to. It does not matter to me at all that these Tf library functions may not be the right functions to use if someone wants to write a file browser or directory traversal or diff utility. So the theoretical scenarios you describe where this change might produce incorrect results don't really sway me. If you can't show me how USD will behave incorrectly you'll have trouble convincing me. So now I'll defend this change from that perspective, understanding that my arguments may be wholly unconvincing if your objective is to allow this function to be used for a wide variety of potential future purposes:

I'm very reluctant to alter this PR (can you tell?) This PR fixed a real bug in production reported by a Houdini user. It has been running in every Houdini USD library for at least 2.5 years. And although I can see the scenarios you describe where this function returns wrong results, I believe that at worst this wrongness will cause the USD library to attempt to read or write a file that it can't read or write. This always has been and always will be an eventuality the USD library must deal with no matter how much pre-checking we do. And in my experience it deals with it quite robustly. The only difference to the user is that the error reported will be "I couldn't write the file s:/path/to/some/file.usd" instead of "I didn't try to write to S:/path/to/some/file.usd because I think it is a bad symbolic link". The latter is not really more useful, and if there is any chance of reporting the latter when in fact an attempt to write the file might succeed, that is a much worse kind of wrong that will have a strong negative impact on users. It is this kind of wrongness that led to the recent removal of all Windows permission pre-checks. Now USD just tries to read or write the file, and it succeeds or it fails (and reports the error). I'm just trying to follow a similar approach here.

Documentation is always good. If you want to point to some docs that could be augmented with this additional detail about reparse points on network drives I'm happy to add something.

As for adding a new `ignoreNetworkReparsePoints=true`, I understand the impulse, but I think it is one that should be resisted. If no code in the USD library sets these parms to "false" in running production code, then it's just untested code serving no purpose. It adds a maintenance burden and provides a pitfall that future users of this function might fall into not knowing the full history of that untested parm. And yes, I know - unit tests. But this is not a test that can be faked. It can only be tested on a Windows test machine with access to a separate Windows-based server with a reparse point set to point to a missing destination. That's a lot of effort and administration to test a parm that no real code ever uses.

Thank you for all your time and deep thinking and detailed responses around this change.

Reply:
Hey Mark - I think we're actually in closer alignment here than you might think.  After posting I had second thoughts about including the whole "potential problems" section, because ultimately, I agree with you - I think this change is a good one, and should go in.  I just wanted to point out some the potential pitfalls, to make sure we were aware of them.

But perhaps it unnecessarily muddies the waters to point those sort of things out, if I'm then just going to say, "...but I still think we should just do it anyway."  I think it made me seem opposed to this PR, which I'm not.  So I apologize for that.

As for `FindFirstFile("somedir\*")` - when I tested it, for existing/accessible directories, it always returns at least one entry, `.`, even if the directory is emtpy.  (And presumably `FindNextFile` would return `..` as well...) That seems to contradict [the docs](https://learn.microsoft.com/en-us/windows/win32/api/fileapi/nf-fileapi-findfirstfilea#return-value), which say that it will return `INVALID_HANDLE_VALUE` and the `GetLastError()` will be `ERROR_FILE_NOT_FOUND (0x2)`, but in my testing I could never see that behavior.  Whenever I tried a "valid" path (local, local junction, remote, or remote junction), it always returned `.`.  And whenever I tried an invalid path (whether local or remote, plain-not-existing or broken-symlink), it always returned `INVALID_HANDLE_VALUE` and `GetLastError()` was `ERROR_PATH_NOT_FOUND (0x3)`.  (You can see the simple CLI tool I made to test it [here](https://github.com/pmolodo/cppFileFunctionTests/blob/b1ef6b4ef827256d06e0264b9f118d8b39dd65d6/findFirstFileMain.cpp).)

And as for documentation - I was just thinking of adding a description of behavior for network paths to the doxygen-parsed comments of [`TfPathExists`](https://github.com/PixarAnimationStudios/OpenUSD/blob/v24.03/pxr/base/tf/fileUtils.h#L42-L44), [`TfIsDir`](https://github.com/PixarAnimationStudios/OpenUSD/blob/v24.03/pxr/base/tf/fileUtils.h#L48-L52), [`TfIsFile`](https://github.com/PixarAnimationStudios/OpenUSD/blob/v24.03/pxr/base/tf/fileUtils.h#L56-L60), and [`TfIsLink`](https://github.com/PixarAnimationStudios/OpenUSD/blob/v24.03/pxr/base/tf/fileUtils.h#L64) in `fileUtils.h`, and to the (non doxygen parsed) comment of [`Tf_HasAttribute`](https://github.com/PixarAnimationStudios/OpenUSD/blob/v24.03/pxr/base/tf/fileUtils.cpp#L66-L67) in `fileUtils.cpp`.  The last wouldn't make it into documentation, per-se, but still seems like it might be helpful there.

And I'm fine with not adding an extra parameter - I was 50-50 on that anyway.

Reply:
A follow up note on the behavior of `FindFirstFile` - I was able to get it to error with `ERROR_FILE_NOT_FOUND (0x2)` if I passed in a glob str that didn't match anything - ie, `FindFirstFile("somedir\nomatch*")`.  So I think it just considers `.` and `..` as "valid matches" for `*`... which... sort of makes sense?

Reply:
I added some comments to the places you suggested, including a mention that in Tf_HasAttribute that we may want to use FindFirstFile at some point to test for existence of the destination, but I did not actually add that indFirstFile code for fear it may not work in all required cases (both files and directories as the targets of Tf_HasAttribute) and platforms (Windows 10, 11, Server?). I don't feel equipped to do this testing, but agree that if this functionality can be sufficiently validated it would be an improvement, perhaps as a separate PR.

Reply:
Thanks! Sounds good to me!
✔️ 💯 👍